src/Uno.Toolkit.UITest/Controls/SafeArea/Given_SafeArea.cs

```diff
@@ -223,36 +225,140 @@ public void When_Inside_Modal()
-			}
-		}
-
-		private void WaitForIsChecked(QueryEx element, bool expected) =>
-			App.WaitForDependencyPropertyValue(element, "IsChecked", expected);
-	}
-}
+		}
+
+		[Test]
+		[TestCase("Soft", "Padding")]
+		[TestCase("Soft", "Margin")]
+		[TestCase("Hard", "Padding")]
+		[TestCase("Hard", "Margin")]
+		[ActivePlatforms(Platform.Android, Platform.iOS)]
+		public void When_SoftInput_Scroll_Into_View(string constraint, string mode)
+		{
+			const string TopTextBox = "TopTextBox";
+			const string BottomTextBox = "BottomTextBox";
+			const string SpacerBorder = "SpacerBorder";
+			var isSoft = string.Equals("Soft", constraint, StringComparison.OrdinalIgnoreCase);
+
+			NavigateToNestedSample("SafeArea_SoftInput_Scroll");
+
+			App.WaitForElement(SpacerBorder, timeoutMessage: $"Timed out waiting for element: {SpacerBorder}");
+
+			var keyboardRect = GetKeyboardRect();
+
+			App.FastTap($"{constraint}ConstraintMode");
+			App.FastTap($"{mode}Mode");
+
+			App.FastTap(TopTextBox);
+
+			App.Wait(TimeSpan.FromSeconds(1));
+
+			// Make sure the keyboard has opened and the focused TextBox is visible above the keyboard
+			Assert.Greater(keyboardRect.Height, 0);
+
+			if (isSoft)
+			{
+				var topRect = GetRectangle(TopTextBox);
+				var bottomRect = GetRectangle(BottomTextBox);
+
+				Assert.False(topRect.IsEmpty);
+				Assert.False(bottomRect.IsEmpty);
+
+				// The focused TextBox should be visible above the keyboard
+				Assert.AreEqual(bottomRect.Bottom, keyboardRect.Top);
+			}
+			else
+			{
+				var topRect = GetRectangle(TopTextBox);
+				var bottomRect = GetRectangle(BottomTextBox);
+
+				Assert.False(topRect.IsEmpty);
+
+				PlatformHelpers.On(
+					iOS: () => Assert.Greater(bottomRect.Top, keyboardRect.Top),
+					Android: () => Assert.True(bottomRect.IsEmpty)
+				);
+			}
+
+			App.DismissKeyboard();
+
+			App.FastTap(BottomTextBox);
+
+			App.Wait(TimeSpan.FromSeconds(1));
+
+			if (isSoft)
+			{
+				var topRect = GetRectangle(TopTextBox);
+				var bottomRect = GetRectangle(BottomTextBox);
+
+				Assert.False(topRect.IsEmpty);
+				Assert.False(bottomRect.IsEmpty);
+
+				Assert.Less(topRect.Bottom, keyboardRect.Top);
+				Assert.AreEqual(bottomRect.Bottom, keyboardRect.Top);
+			}
+			else
+			{
+				App.WaitForNoElement(TopTextBox, timeoutMessage: $"Timed out waiting for no element: {TopTextBox}");
+				var bottomRect = GetRectangle(BottomTextBox);
+
+				Assert.False(bottomRect.IsEmpty);
+				Assert.AreEqual(bottomRect.Bottom, keyboardRect.Top);
+			}
+
+			Rectangle GetKeyboardRect()
+			{
+				var initialBottom = App.GetPhysicalRect(BottomTextBox).Bottom;
+
+				App.FastTap($"SoftConstraintMode");
+				App.FastTap(TopTextBox);
+				App.Wait(TimeSpan.FromSeconds(1));
+
+				var newBottom = App.GetPhysicalRect(BottomTextBox).Bottom;
+				var keyboardHeight = initialBottom - newBottom;
+				var keyboardTop = initialBottom - keyboardHeight;
+				var windowRect = App.GetPhysicalScreenDimensions();
+
+				App.DismissKeyboard();

```

Comment:
should move out of GetKeyboardRect()

Reply:
Not sure I follow, this method is used to calculate what the height of the keyboard will be so it can be used for the rest of the test. Once I get my values I close the keyboard so we are back to the same state that we were at when we entered the method

Reply:
you are inside a method name `GetKeyboardRect()`, i dont expect side-effect from it
if you are done with the kb afterward, then be explicit about it:
```
//BAD
var kb = GetKeyboardRect(); // oh btw, this also magically make the keyboard vanish

//GOOD
var kb = GetKeyboardRect();
App.DismissKeyboard();
```

Reply:
if you didn't have the keyboard open before calling this, wouldn't GetKeyboardRect() also have a side-effect of opening the keyboard and leaving it opened once it completes?